File: app/src/main/java/de/tum/in/tumcampusapp/models/DeviceRegister.java
package de.tum.in.tumcampusapp.models;

import android.content.Context;

import java.math.BigInteger;
import java.security.SecureRandom;
import java.util.Date;

import de.tum.in.tumcampusapp.auxiliary.AuthenticationManager;
import de.tum.in.tumcampusapp.exceptions.NoPrivateKey;

public class DeviceRegister {

    private String signature;
    private String date;
    private String rand;
    private String device;
    private String publicKey;
    private String member = "";

    public DeviceRegister(Context c, String publickey, ChatMember member) throws NoPrivateKey {
        //Create some data
        this.date = (new Date()).toString();
        this.rand = new BigInteger(130, new SecureRandom()).toString(32);
        this.device = AuthenticationManager.getDeviceID(c);
        this.publicKey = publickey;
        if (member != null) {
            this.member = member.getLrzId();
        }
        //Sign this data for verification
        AuthenticationManager am = new AuthenticationManager(c);
        this.signature = am.sign(date + rand + this.device + this.member);
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/models/tumcabe/DeviceRegister.kt
package de.tum.`in`.tumcampusapp.models.tumcabe

import android.content.Context
import de.tum.`in`.tumcampusapp.auxiliary.AuthenticationManager
import de.tum.`in`.tumcampusapp.exceptions.NoPrivateKey
import java.math.BigInteger
import java.security.SecureRandom
import java.util.*

data class DeviceRegister(var signature: String = "",
                          var date: String = "",
                          var rand: String = "",
                          var device: String = "",
                          var publicKey: String = "",
                          var member: ChatMember? = null) {
    companion object {
        @Throws(NoPrivateKey::class)
        fun getDeviceRegister(c: Context, publickey: String, member: ChatMember?): DeviceRegister {
            val am = AuthenticationManager(c)
            val date = Date().toString()
            val rand = BigInteger(130, SecureRandom()).toString(32)
            val device = AuthenticationManager.getDeviceID(c)

            return DeviceRegister(date = date, rand = rand, device = device, publicKey = publickey, member = member,
                    signature =
                    if (member == null)
                        am.sign(date + rand + device)
                    else
                        am.sign(date + rand + device + member.lrzId + member.id)
            )
        }
    }
}
